swarm/storage/mru/resource.go

```diff
@@ -525,63 +558,73 @@ func (h *Handler) Load(addr storage.Address) (*resource, error) {
 
 	// retrieve metadata from chunk data and check that it matches this mutable resource
-	signature, period, version, name, data, multihash, err := h.parseUpdate(chunk.SData)
-	if rsrc.name != name {
-		return nil, NewError(ErrNothingToReturn, fmt.Sprintf("Update belongs to '%s', but have '%s'", name, rsrc.name))
-	}
-	log.Trace("resource index update", "name", rsrc.name, "namehash", rsrc.nameHash, "updatekey", chunk.Addr, "period", period, "version", version)
+	mru, err := parseUpdate(chunk.SData)
+	log.Trace("resource index update", "name", rsrc.name, "updatekey", chunk.Addr, "period", mru.period, "version", mru.version)
 
-	// check signature (if signer algorithm is present)
 	// \TODO maybe this check is redundant if also checked upon retrieval of chunk
-	if signature != nil {
-		digest := h.keyDataHash(chunk.Addr, data)
-		_, err = getAddressFromDataSig(digest, *signature)
-		if err != nil {
-			return nil, NewError(ErrUnauthorized, fmt.Sprintf("Invalid signature: %v", err))
-		}
+
+	ownerAddr, err := mru.Verify(chunk.Addr)
+	if err != nil {
+		return nil, NewError(ErrUnauthorized, fmt.Sprintf("Invalid signature: %v", err))
+	}
+
+	// Check if who signed the resource update really owns the resource
+	if !verifyResourceOwnership(ownerAddr, mru.metaHash, mru.rootAddr) {
+		return nil, NewError(ErrUnauthorized, fmt.Sprintf("signature is valid but signer does not own the resource: %v", err))
 	}
 
 	// update our rsrcs entry map
 	rsrc.lastKey = chunk.Addr
-	rsrc.lastPeriod = period
-	rsrc.version = version
+	rsrc.period = mru.period
+	rsrc.version = mru.version
 	rsrc.updated = time.Now()
-	rsrc.data = make([]byte, len(data))
-	rsrc.Multihash = multihash
+	rsrc.data = make([]byte, len(mru.data))
+	rsrc.multihash = mru.multihash
 	rsrc.Reader = bytes.NewReader(rsrc.data)
-	copy(rsrc.data, data)
-	log.Debug(" synced", "name", rsrc.name, "key", chunk.Addr, "period", rsrc.lastPeriod, "version", rsrc.version)
-	h.set(chunk.Addr.Hex(), rsrc)
+	copy(rsrc.data, mru.data)
+	log.Debug(" synced", "name", rsrc.name, "key", chunk.Addr, "period", rsrc.period, "version", rsrc.version)
+	h.set(chunk.Addr, rsrc)
 	return rsrc, nil
 }
 
 // retrieve update metadata from chunk data
 // mirrors newUpdateChunk()
-func (h *Handler) parseUpdate(chunkdata []byte) (*Signature, uint32, uint32, string, []byte, bool, error) {
+func parseUpdate(chunkdata []byte) (*SignedResourceUpdate, error) {
 	// absolute minimum an update chunk can contain:
 	// 14 = header + one byte of name + one byte of data
+
+	// 2 bytes header Length
+	// 2 bytes data length
+	// 4 bytes period
+	// 4 bytes version
+	// 32 bytes rootAddr reference
+	// 32 bytes metaHash digest
+
 	if len(chunkdata) < 14 {
-		return nil, 0, 0, "", nil, false, NewError(ErrNothingToReturn, "chunk less than 13 bytes cannot be a resource update chunk")
+		return nil, NewError(ErrNothingToReturn, "chunk less than 13 bytes cannot be a resource update chunk")
```

Comment:
14

Reply:
This is all refactored in #704 and those lone numbers gone.